src/libraries/System.Text.Json/System.Text.Json.SourceGeneration/JsonSourceGeneratorHelper.cs

```diff
@@ -0,0 +1,550 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.CodeAnalysis;
+
+namespace System.Text.Json.SourceGeneration
+{
+    internal sealed class JsonSourceGeneratorHelper
+    {
+        // Simple handled types with typeinfo.
+        private static readonly HashSet<Type> s_simpleTypes = new HashSet<Type>
+        {
+            typeof(bool),
+            typeof(int),
+            typeof(double),
+            typeof(long),
+            typeof(string),
+            typeof(char),
+            typeof(DateTime),
+            typeof(DateTimeOffset),
+        };
+
+        // Generation namespace for source generation code.
+        private string _generationNamespace;
+
+        // Full assembly type name for key and a generated-source for value.
+        public Dictionary<Type, string> _types = new Dictionary<Type, string>();
+
+        // Contains name of types that failed to be generated.
+        private HashSet<Type> _failedTypes = new HashSet<Type>();
+        public List<Type> FailedTypes { get { return _failedTypes.ToList(); } }
+
+        // Contains list of diagnostics for the code generator.
+        private List<Diagnostic> _diagnostics = new List<Diagnostic>();
+        public List<Diagnostic> Diagnostics { get { return _diagnostics; } }
+
+        // Diagnostic descriptors for user.
+        private DiagnosticDescriptor _initiatingTypeClass;
+        private DiagnosticDescriptor _failedToGenerateTypeClass;
+        private DiagnosticDescriptor _failedToAddNewTypesFromMembers;
+
+        public JsonSourceGeneratorHelper(string generationNamespace)
+        {
+            _generationNamespace = generationNamespace;
+
+            // Initiate diagnostic descriptors.
```

Comment:
Is there test coverage for when these descriptors are fired?

Reply:
Added